atecc508a_config_dev.h

```diff
@@ -0,0 +1,182 @@
+   * (8-11)  0000   WriteKey - Irrelevant, since clear writes are permitted.
+   * (7)     1      IsSecret - The contents of this slot are secret.
+   * (6)     0      EncryptRead - Reads from this slot are prohibited.
+   * (5)     1      LimitedUse - Limited use according to the unique
+   *                             slot 15 rules.
+   * (4)     0      NoMac - The key stored in the slot can be used by
+   *                        all commands.
+   * (0-3)   0111   ReadKey - (priv key in slot) - External and internal
+   *                          signatures are enabled, ECDH operations too. */
+  0xA7, 0x20, /* 50-51 SlotConfig 15 */
+
+  /* Monotonic counter that can be connected to keys to determine how many
+   * times a key can be used. 0x00 means no limit. */
+  0xFF, 0xFF, 0xFF, 0xFF, 0x00, 0x00, 0x00, 0x00, /* 52-59 Counter[0] */
+  0xFF, 0xFF, 0xFF, 0xFF, 0x00, 0x00, 0x00, 0x00, /* 60-67 Counter[1] */
+
+  /* LastKeyUse - 128 bits controlling limited use of KeyID 15, initialized to 0xFF. */
+  0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+  0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, /* 68-83 LastKeyUse */
+
+  0x00, /* 84 UserExtra - one byte value that can be modified via UpdateExtra calls. */
+  0x00, /* 85 Selector  - which device will remain active after a Pause call.        */
+
+  0x55, /* 86 LockValue - OTP zone unlocked (0x00 to lock). */
+  0x55, /* 87 LockConfig - config zone unlocked (value needed for
+         *                 correct CRC calculation). */
+  0xFF, 0xFF, /* 88-89 SlotLocked - one bit for each slot (0 means locked, intuitively). */
+  0x00, 0x00, /* 90-91 RFU - must be zero. */
+  0x00, 0x00, 0x00, 0x00, /* 92-95 X509format - 0 means ignore formatting restrictions. */
+
+  /* 96-127 - KeyConfig - usage permissions and control, two bytes per slot. */
+  /* 0x1300 - ‭0001 0011 0000 0000‬
+   * (15-14) 00     X509id - Public key validation by any format signature by parent;
+   * (13)    0      RFU - Must be zero.
+   * (12)    0      IntrusionDisable - Use of key independent of the state of IntrusionLatch.
+   * (11-8)  0000   AuthKey - Zero becaues ReqAuth is zero.
+   * (7)     0      ReqAuth - No prior authorization is required before using the key.
+   * (6)     0      ReqRandom - A random nonce is not required for a specific group of commands.
+   * (5)     0      Lockable - Slot cannot be individually locked using the Lock command.
+   * (4-2)   100    KeyType - P256 NIST ECC key
+   * (1)     1      PubInfo - Public version of a key can always be generated.
+   * (0)     1      Private - Contains a private key */
+  0x13, 0x00, /*  96-97  KeyConfig slot 0 */
+  0x13, 0x00, /*  98-99  KeyConfig slot 1 */
+  0x13, 0x00, /* 100-101 KeyConfig slot 2 */
+  0x13, 0x00, /* 102-103 KeyConfig slot 3 */
+  0x13, 0x00, /* 104-105 KeyConfig slot 4 */
+  0x13, 0x00, /* 106-107 KeyConfig slot 5 */
+  0x13, 0x00, /* 108-109 KeyConfig slot 6 */
+  0x13, 0x00, /* 110-111 KeyConfig slot 7 */
```

Comment:
You might be right - I assumed that it's a key (or should be treated as one), but that's probably due to my lack of knowledge.
From the KeyType documentation:
`If the slot contains an ECC public or private key, then the key type field below must be set
to 0b100. If the slot contains any other kind of data, key, or secret, then this field must be
set to 0b111 for proper operation.
100 = P256 NIST ECC key
111 = Not an ECC key`

Reply:
We probably have to try this out to see if we can make use of Slot 7 after it gets a secret placed there by a Slot 6 operation. It doesn't help to provision these devices for this use just to have it not work in practice as the key type field is incorrect.

My reading of the datasheet leads me to believe that the key type should be `0b111`, but datasheets (or our interpretations of them) can be wrong. It's best to try it out and establish some ground truth.

Considering that the 508A can't do AES or any PRF (e.g. the TLS 1.2 PRF would be useful), we'd need to be able to read the secret to continue with a TLS operation after the initial ECDH. We may want to do this using the 508A's encrypted read feature, or trust that the I2C transport security is good enough.

Reply:
There are three ways to proceed:

- Treat the master secret as a key (is it a key?) and see if we can encrypt/decrypt/generate public/sign with it - is it a valid use case for the master secret?
- Treat it as raw data and change the config of the slot to be secret, but not a key and enable encrypted reads from the slot;
- Treat it as raw data and change the config of the slot to enable unencrypted raw data readout.

Which one would you prefer, @Patater?

Reply:
Option 1 doesn't make sense. Option 3 seems good, but I wonder if we can't find an application note about using this chip with TLS perhaps and see how they recommend we use it.

Reply:
@Patater - @yanesca mentioned that option 1 makes the most sense for him, and this makes the current version of the PR feasible.

Reply:
@AndrzejKurek I also mentioned that I think that the data sheet clearly states that here the type must be `0b111`. I think we should still fix this in this PR.

Reply:
Option 1 says `0b100`, which is what doesn't make sense to me. Moving to `0b111` makes sense to me.

Reply:
Then I might have misinterpreted the options. Here is what I meant to say:
- Set the type to `0b111`
- Configure it as a secret
- Enabling encrypted or unencrypted read should not be necessary

Reply:
I would rather enable encrypted reads, since it makes the configuration more permissive, and it's a developer's configuration.

Reply:
I suggested to disable it, because based on my current understanding I don't see how that would provide any value in the TLS 1.2 PRF case.

But I think you are right, it would probably be better to enable encrypted reads: we only will learn more about the SE over time, maybe later we find a way to get value out of this option or maybe we find some use for it in a non-TLS use case.

Reply:
After going through the spec - private keys cannot be read in any way, not even by the encrypted read. Slot 7 is already configured as secret and private, so the only change to be made is changing its KeyType to 0b111.

Reply:
Updated.

Reply:
I am sorry I am not sure I understand. Will we be able to do encrypted read on Slot 7 if we do that?

Reply:
No, we will not, but I assumed we don't need to since 

> Enabling encrypted or unencrypted read should not be necessary

Reply:
Yes, but as you pointed out it is a very black and white approach and for development purposes enabling an encrypted read would be a better choice.

If we declare it as not a key, will we be able to use it as a key in HMAC? (To be specific, in the terminology of the data sheet I mean using the SHA command in HMAC mode and not the HMAC command.)

Reply:
The only restriction that I have found was the "NoMac" field of SlotConfig, if set to 1:

> 1 = The key stored in the slot is intended for verification usage and cannot be used by the MAC or HMAC commands. When this key is used to generate or modify TempKey, then that value may not be used by the MAC and HMAC commands. Also cannot be used with the SHA command in HMAC mode.

Description of the SHA command:
>1. HMACstart: Initialization of the HMAC calculation engine and initialization of the SHA context in memory. A stored key legal for use with SHA operations must be specified. 

Description of the HMACStart mode parameter:
>100 (HMACstart) = Load TempKey with the initialization value for SHA2-56. Length field specifies the key to be used for the HMAC calculation. No message bytes are accepted.

Description of the second parameter:
>KeyID for the HMAC key if Mode is “HMACstart”.

So although the datasheet refers to the parameter as a key, designated by providing `KeyID`, there's no clear indication that the slot has to have they KeyType slot set to the key value.
From the KeyType description:

> The four ECC commands that use ECC keys (i.e. GenKey, Sign, ECDH, and Verify) will operate only on data slots in which this field is set to one of the legal ECC key types. Any attempt to use any SHA-256 computation commands (i.e. CheckMac, DeriveKey, MAC, or HMAC) on a slot configured to be an ECC private key will result in an error.
Keys that will be the source or destination of the SHA-256 computation commands (i.e. CheckMac, DeriveKey, MAC, or HMAC) should be set to a KeyType of 0b111. Proper operation of the device is not guaranteed if these commands are attempted with any other KeyType. The GenDig command may operate on any slot type other than for ECC private keys.

Reply:
I am sorry, I just realised that the chip does not have AES.    
                                                                                
How I imagined this working with TLS:                                           
                                                                                
1. Use Slot 6 ECDH to generate the shared secret (pre-master secret)            
2. Use Slot 7 HMAC to generate the master secret and put it to Slot X           
3. Use Slot X HMAC to generate the transport secrets and put them into Slot Y   
4. Use Slot Y to en/decrypt transport traffic                                   
                                                                                
Unfortunately this is 508a, the chip does not have AES (at least I couldn't     
find it in the data sheet) and the transport secrets need to be read out to     
perform encryption in software. This read out of course would be best done
with encrypted readout.                                                
                                                                                
This raises a couple of questions that could affect the slot configuration:                                                                        
- One option is to just read out the key and do encryption in software in step 4. In this case can we reuse Slot 7 in step 2 or do we need to prepare another slot that can be used as a HMAC key?                                            
- Or we could read out Slot 7 and use the SHA command with keys stored in memory
  in steps 2-4. Can the chip do that?